README.md
# NXP Secure Provisioning SDK

**Secure Provisioning SDK (SPSDK)** enables connection and communication with target devices for purposes of secure provisioning and programming. Delivered as python library with command-line applications for direct utilization.

<img src="docs/_static/images/spsdk-architecture.png" alt="drawing" width="600"/>

## Internal Link

* [Documentation](http://spsdk.nxp.com/)
* [BitBucket](https://bitbucket.sw.nxp.com/projects/SPSDK/repos/spsdk/browse)
* [Confluence](https://confluence.sw.nxp.com/display/SPSDK)
* [SharePoint](https://nxp1.sharepoint.com/sites/BSDKCommunity)

## External Link

* [GitHub](https://github.com/NXPmicro/spsdk)
* [PyPi](https://pypi.org/project/spsdk/)
* [Documentation](https://spsdk.readthedocs.io)

## Supported Devices

Following NXP devices are supported:

- LPC55 [S6x](https://www.nxp.com/products/processors-and-microcontrollers/arm-microcontrollers/general-purpose-mcus/lpc5500-cortex-m33/high-efficiency-arm-cortex-m33-based-microcontroller-family:LPC55S6x) / [S3x](https://www.nxp.com/products/processors-and-microcontrollers/arm-microcontrollers/general-purpose-mcus/lpc5500-cortex-m33:LPC5500_SERIES) / [S2x](https://www.nxp.com/products/processors-and-microcontrollers/arm-microcontrollers/general-purpose-mcus/lpc5500-cortex-m33/lpc552x-s2x-mainstream-arm-cortex-m33-based-microcontroller-family:LPC552x-S2x) / [S1x](https://www.nxp.com/products/processors-and-microcontrollers/arm-microcontrollers/general-purpose-mcus/lpc5500-cortex-m33/lpc551x-s1x-baseline-arm-cortex-m33-based-microcontroller-family:LPC551X-S1X) / [S0x](https://www.nxp.com/products/processors-and-microcontrollers/arm-microcontrollers/general-purpose-mcus/lpc5500-cortex-m33/lpc550x-s0x-baseline-arm-cortex-m33-based-microcontroller-family:LPC550x)
- i.MX RT [600](https://www.nxp.com/products/processors-and-microcontrollers/arm-microcontrollers/i-mx-rt-crossover-mcus/i-mx-rt600-crossover-mcu-with-arm-cortex-m33-and-dsp-cores:i.MX-RT600) / [500](https://www.nxp.com/products/processors-and-microcontrollers/arm-microcontrollers/i-mx-rt-crossover-mcus/i-mx-rt500-crossover-mcu-with-arm-cortex-m33-core:i.MX-RT500)
- i.MX RT [1064](https://www.nxp.com/products/processors-and-microcontrollers/arm-microcontrollers/i-mx-rt-crossover-mcus/i-mx-rt1064-crossover-mcu-with-arm-cortex-m7-core:i.MX-RT1064) / [1060](https://www.nxp.com/products/processors-and-microcontrollers/arm-microcontrollers/i-mx-rt-crossover-mcus/i-mx-rt1060-crossover-mcu-with-arm-cortex-m7-core:i.MX-RT1060) / [1050](https://www.nxp.com/products/processors-and-microcontrollers/arm-microcontrollers/i-mx-rt-crossover-mcus/i-mx-rt1050-crossover-mcu-with-arm-cortex-m7-core:i.MX-RT1050) / [1020](https://www.nxp.com/products/processors-and-microcontrollers/arm-microcontrollers/i-mx-rt-crossover-mcus/i-mx-rt1020-crossover-mcu-with-arm-cortex-m7-core:i.MX-RT1020) / [1010](https://www.nxp.com/products/processors-and-microcontrollers/arm-microcontrollers/i-mx-rt-crossover-mcus/i-mx-rt1010-crossover-mcu-with-arm-cortex-m7-core:i.MX-RT1010)
- i.MX RT [1170](https://www.nxp.com/products/processors-and-microcontrollers/arm-microcontrollers/i-mx-rt-crossover-mcus/i-mx-rt1170-crossover-mcu-family-first-ghz-mcu-with-arm-cortex-m7-and-cortex-m4-cores:i.MX-RT1170) / [1160](https://www.nxp.com/design/development-boards/i-mx-evaluation-and-development-boards/i-mx-rt1160-evaluation-kit:MIMXRT1160-EVK)

## Supported environments

- Windows 10, 64bit
- Ubuntu 18.04 or above, 64bit
- Mac OS 10.15 or above, x64, ARM64

## Installation

- Make sure to have [Python](https://www.python.org) 3.7+ installed
- Create a virtual environment (venv, pipenv, etc.)

> SPSDK is tested on Python >=3.7 and <3.11 interpreter, old version 2.x is not supported

### BitBucket

``` bash
    $ pip install -U https://bitbucket.sw.nxp.com/rest/api/latest/projects/spsdk/repos/spsdk/archive?format=zip
```
> If you will be asked for credentials, use your NXP login and password

### SPSDK from sources - BitBucket:

``` bash
    $ git clone ssh://git@bitbucket.sw.nxp.com/spsdk/spsdk.git
    $ cd spsdk
    $ pip install -r requirements-develop.txt
    $ pip install -U -e .
```

> In **Windows OS** you need to install [Microsoft Visual C++ Build Tools](https://www.scivision.dev/python-windows-visual-c-14-required/)

 Note: In case of problems during installation, please make sure that you have the latest pip version.
 You can upgrade pip using this command: 'pip install --upgrade pip'.

## Usage

- See [examples](examples) directory
- See [application](spsdk/apps) directory

---
**i.Mx RT 1050**

To run examples using i.MX RT 1050 you need to download a flashloader:
- Go to: https://www.nxp.com/webapp/sps/download/license.jsp?colCode=IMX-RT1050-FLASHLOADER
- Review the license agreement, download and unzip the package
- Convert the elf file into bin (For this operation you need to have MCUXpresso IDE, IAR or Keil)
  - run ```python tools\flashloader_converter.py --elf-path <path/to/flashloader.elf> --ide-type <mcux | iar | keil> --ide-path <path/to/IDE/install/folder```

---

## Dependencies

The core dependencies are included in [requirements.txt](requirements.txt).

The dependencies for the development and testing are included in [requirements-develop.txt](requirements-develop.txt).


release_notes.txt
Release Notes for Secure Provisioning SDK
==========================================
Secure Provisioning SDK (SPSDK) is a unified, reliable, and easy-to-use SW library. It targets a wide
portfolio of NXP MCUs, providing a solid foundation from quick customer prototyping up to
production deployment. The library allows the user to connect and communicate with a device,
configure the device, prepare, download and upload data, including security operations. It is
delivered in the form of a python library and command-line applications.


Version: 1.10.1
================
Date: 26-May-2023

New features
- tphost/tpconfig:
    - add support for LPC55S3x
- nxpimage:
    - add possibility to define multiple regions in OTFAD in one data blob

Supported devices
=================

- i.MX RT1170, RT1160
- i.MX RT1064, RT1060, RT1050, RT1020, RT1010
- i.MX RT685S, RT595S
- LPC55S6x, LPC55S3x, LPC55S2x, LPC55S1x, LPC55S0x

Note: More details in devices.txt

System Requirements
===================

- Windows 10, 64-bit
- Ubuntu 18.04 or above, 64-bit
- Mac OS 10.15 or above, x64, ARM64

Supported Environment
=====================

Python 3.8+ interpreter, old version 2.x is not supported


Revision History
================
1.10.1
- nxpimage:
    - support encrypted image hab
    - support for RT11xx and RT10xx
    - improve OTFAD/IEE names generation
    - fix XMCD load_from_config
    - fix IEE template
- add API to retrieve info about fuses
- fix circular dependency in signature provider import
- fix issue with loading keys as INT
- not enable logging when spsdk is used as a library
1.10.0
- blhost:
    - add new command: ele_message
- ifr:
    -  add commands read/write
- nxpdebugmbox:
    - add command: read UUID from device
    - update PyOCD to latest version to support CMSIS DAP FW v3
    - fix AP selection issue for PyOCD and PEMICRO
    - fix DAC verification when there is only 1 root key
- nxpdevhsm:
    - USER_PCK rename to CUST_MK_SK
- nxpimage:
    - add subcommand group for generate and parse certificate block
    - replace private key to signature provider in master boot image
    - OTFAD support for RT1170
    - fix MBI issue with HMAC
- pfr:
    - add CMPA erase command
- shadowregs:
    - fix endianness for OTP MASTER KEY
- drop support for Python 3.7
1.9.1
- nxpdebugmbox:
    - fix Linux error on PyOCD
    - fix PyOCD and PEmicro connection for kw45xx and k32w1xx
- nxpdevhsm:
    - fix buffer base address for DevHSM operations
    - split reset option in nxpdevhsm into two; disable init reset by default
- nxpimage:
    - fix handling exception when the root cert index is wrong
- tphost/tpconfig:
    - incorrect output in TP PG command in case of an failure
1.9.0
- nxpdebugmbox:
    - add check of root of trust hash in dat authentication
    - enable debug authentication protocol on RT1180
    - remove duplicated option --protocol for gendc command
- nxpdevhsm:
    - reset target before and after DevHSM SB3 file creation
    - fix skipping commands from config file
- nxpimage:
    - XMCD support
    - signed messages support for RT1180
    - add bootable image for RT10xx, RT1180, RT1170, LPC55S3x
    - implement IEE encryption
    - support Memory ID for erase in sb21
    - support Memory ID for enable and load in sb21
    - implement JUMP and JUMP_SP commands in BD file for SB2.1
    - enable encryption in AHAB container
    - fix non working 384/521 ECC keys for signature in AHAB container
    - fix CRC mode in external flash for lpc55s3x
    - failure on start due to boot_image hook definition
- pfr:
    - command line parameter '-t' is duplicated
- tphost/tpconfig:
    - create command for loading ProvFW
    - add command for retrieving TP_RESPONSE without models or smart card
    - smart card reader name hash identification
    - TPhost load-tpfw requires TP device definition
    - OEM ProvFW boot-check incorrectly fails with non-verbose flavor
- debug authentication improvements
- unify memory access cross all debuggers
- replace json file with yml file for TZ
- support for k32w1xx, kw45xx
- improve format of debugging logger

1.8.0
- nxpdevscan:
    - fix hanging up for serial communication
- nxpimage:
    - add support for BEE
    - enable OTFAD on RT1180
- pfr:
    - move the functionality of pfrc tool into PFR tool
- tphost/tpconfig:
    - implement USB re-enumeration in TPHost after OEM ProvFW is started
    - create command for checking the Chain of Trust used in TP
    - investigate TP performance loss during device reset after TP is completed
    - add possibility to select TP SmartCard via card reader's name
    - blhost_port should not be mandatory in TP target settings
    - fix disabling timeout in TP is ignored
- unify option for getting template across tools
- add API for parsing XMCD
- support cryptography >= 37.0.0
- support bincopy 17.14
- fix documentation regarding SB31 programFuses

1.7.1
- nxpdebugmbox:
    - fix debug authentication for RT595
- nxpimage:
    - add OTFAD support for RT5xx and RT6xx devices
    - fix sb21 command line argument in documentation
- pfr:
    - read command allows independent binary and yaml exports
- shadowregs:
    - new subcommand: fuses-script
- fix the use of pyyaml's load in tests (use safe_load())
- add OEM cert size check into TPConfig

1.7.0
- nxpimage application as replacement for elftosb
- nxpcrypto application for generating and verifying keys, certificates, hash digest, converting key's format
- trust provisioning applications (tphost and tpconfig)
- blhost:
    - support LifeCycleUpdate command for RT1180
    - add option to specify peripheral index of SPI/I2C for LIBUSBSIO
    - allow lowercase names in the filter for USB mboot devices
    - efuse_program_once returns failure message when using 'lock' option but still the fuse is burnt
    - fix in re-scanning LIBUSBSIO devices when target MCU is not connected
    - scan_usb() should return nxp devices

setup2.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-
#
# Copyright 2019-2020 NXP
#
# SPDX-License-Identifier: BSD-3-Clause

import os
import sys
from typing import List

import pip  # type: ignore

from setuptools import setup, find_packages  # type: ignore


def sanitize_version(version_str: str) -> str:
    """Change the version into string relevant for comparison."""
    sanitizer = {
        4: version_str + '.0',
        5: '0' + version_str,
        6: version_str
    }
    return sanitizer[len(version_str)]


if sys.version_info >= (3, 8, 0) and sanitize_version(pip.__version__) < '19.2.3':
    print(f"With python {sys.version}, you're using an old version of pip: {pip.__version__}")
    print("Please update pip using: 'python -m pip install --upgrade pip'.")
    sys.exit(1)


def get_requirements() -> List[str]:
    """Get the list of requirements from requirements.txt file."""
    with open('requirements-readthedocs.txt') as req_file:
        requirements = req_file.read().splitlines()
    return [x for x in requirements if "astunparse" not in x] + ["astunparse"]


with open("README.md", "r") as f:
    long_description = f.read()

# extract version info indirectly
version_info = {}   # type: ignore
base_dir = os.path.dirname(__file__)
with open(os.path.join(base_dir, "spsdk", "__version__.py")) as f:
    exec(f.read(), version_info)

setup(
    name='spsdk',
    version=version_info["__version__"],
    description='Open Source Secure Provisioning SDK for NXP MCU/MPU',
    url="https://github.com/NXPmicro/spsdk",
    author="NXP",
    author_email="michal.starecek@nxp.com",
    license='BSD-3-Clause',
    long_description=long_description,
    long_description_content_type="text/markdown",
    platforms="Windows, Linux, Mac OSX",
    python_requires=">=3.6",
    setup_requires=[
        'setuptools>=40.0'
    ],
    install_requires=[
        get_requirements(),
        'astunparse @ git+https://github.com/tbennun/astunparse#egg=astunparse'
    ],
    include_package_data=True,
    classifiers=[
        'Development Status :: 3 - Alpha',
        'Programming Language :: Python :: 3 :: Only',
        'Programming Language :: Python :: 3.6',
        'Programming Language :: Python :: 3.7',
        'Programming Language :: Python :: 3.8',
        'Operating System :: POSIX :: Linux',
        'Operating System :: Microsoft :: Windows',
        'Operating System :: MacOS :: MacOS X',
        'License :: OSI Approved :: BSD License',
        'Topic :: Scientific/Engineering',
        'Topic :: Software Development :: Embedded Systems',
        'Topic :: System :: Hardware',
        'Topic :: Utilities'
    ],
    packages=find_packages(exclude=["tests.*", "tests", "examples"]),
    entry_points={
        'console_scripts': [
            'elftosb=spsdk.apps.elftosb:safe_main',
            'pfr=spsdk.apps.pfr:safe_main',
            'pfrc=spsdk.apps.pfrc:safe_main',
            'blhost=spsdk.apps.blhost:safe_main',
            'sdphost=spsdk.apps.sdphost:safe_main',
            'sdpshost=spsdk.apps.sdpshost:safe_main',
            'spsdk=spsdk.apps.spsdk_apps:safe_main',
            'nxpkeygen=spsdk.apps.nxpkeygen:safe_main',
            'nxpdebugmbox=spsdk.apps.nxpdebugmbox:safe_main'
        ],
    },
)
